packages/puppeteer-core/src/bidi/Page.ts

```diff
@@ -774,6 +775,136 @@ export class BidiPage extends Page {
     });
   }
 
+  override async cookies(...urls: string[]): Promise<Cookie[]> {
+    const normalizedUrls = (urls.length ? urls : [this.url()]).map(url => {
+      return new URL(url);
+    });
+
+    const bidiCookies = await this.#connection.send('storage.getCookies', {
+      partition: {
+        type: 'context',
+        context: this.mainFrame()._id,
+      },
+    });
+    return bidiCookies.result.cookies
+      .map(c => {
+        return this.#bidiToPuppeteerCookie(c);
+      })
+      .filter(c => {
+        return normalizedUrls.some(u => {
+          return BidiPage.#testUrlMatchCookie(c, u);
+        });
+      });
+  }
+
+  /**
+   * Check domains match.
+   * According to cookies spec, this check should match subdomains as well, but CDP
+   * implementation does not do that, so this method matches only the exact domains, not
+   * what is written in the spec:
+   * https://datatracker.ietf.org/doc/html/rfc6265#section-5.1.3
+   */
+  static #testUrlMatchCookieHostname(
+    cookie: Cookie,
+    normalizedUrl: URL
+  ): boolean {
+    const cookieDomain = cookie.domain.toLowerCase();
+    const urlHostname = normalizedUrl.hostname.toLowerCase();
+    return cookieDomain === urlHostname;
+  }
+
+  /**
+   * Check paths match.
+   * Spec: https://datatracker.ietf.org/doc/html/rfc6265#section-5.1.4
+   */
+  static #testUrlMatchCookiePath(cookie: Cookie, normalizedUrl: URL): boolean {
+    const uriPath = normalizedUrl.pathname;
+    const cookiePath = cookie.path;
+
+    if (uriPath === cookiePath) {
+      // The cookie-path and the request-path are identical.
+      return true;
+    }
+    if (uriPath.startsWith(cookiePath)) {
+      // The cookie-path is a prefix of the request-path.
+      if (cookiePath.endsWith('/')) {
+        // The last character of the cookie-path is %x2F ("/").
+        return true;
+      }
+      if (uriPath[cookiePath.length] === '/') {
+        // The first character of the request-path that is not included in the cookie-path
+        // is a %x2F ("/") character.
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Checks the cookie matches the URL according to the spec:
+   */
+  static #testUrlMatchCookie(cookie: Cookie, url: URL): boolean {
+    const normalizedUrl = new URL(url);
+    console.assert(cookie !== undefined);
+    if (!this.#testUrlMatchCookieHostname(cookie, normalizedUrl)) {
+      return false;
+    }
+    return this.#testUrlMatchCookiePath(cookie, normalizedUrl);
+  }
+
+  #bidiToPuppeteerCookie(bidiCookie: Bidi.Network.Cookie): Cookie {
```

Comment:
let's keep the bidi to Puppeteer mapping functions in this file but make it a module-level function (it does not use any of the page instance state)

Reply:
done